package/gluon-statistics-mcast/src/respondd.c

```diff
@@ -0,0 +1,478 @@
+  modification, are permitted provided that the following conditions are met:
+
+    1. Redistributions of source code must retain the above copyright notice,
+       this list of conditions and the following disclaimer.
+    2. Redistributions in binary form must reproduce the above copyright notice,
+       this list of conditions and the following disclaimer in the documentation
+       and/or other materials provided with the distribution.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+
+#include <respondd.h>
+
+#include <json-c/json.h>
+#include <libgluonutil.h>
+
+#include <dirent.h>
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <unistd.h>
+
+#define BPFCOUNTD_SOCKS_DIR "/var/run/bpfcountd/gluon-sockets"
+
+enum capdir {
+	CAP_IN,
+	CAP_OUT,
+	__CAP_COUNT,
+};
+
+static const char *dirstr[__CAP_COUNT] = {
+	[CAP_IN] = "in",
+	[CAP_OUT] = "out",
+};
+
+static int check_socket_suffix(char *socketname)
+{
+	unsigned int offset;
+
+	if (strlen(socketname) <= strlen(".sock"))
+		return -EINVAL;
+
+	offset = strlen(socketname) - strlen(".sock");
+	if (strcmp(&socketname[offset], ".sock"))
+		return -EINVAL;
+
+	return 0;
+}
+
+static int parse_socket_dir(char *socketname, enum capdir *dir)
+{
+	char *dir_start, *dir_end;
+
+	dir_end = strrchr(socketname, '.');
+	if (!dir_end)
+		return -EINVAL;
+
+	/* check for .in */
+	dir_start = dir_end - strlen(".in");
+	if (dir_start < socketname)
+		return -EINVAL;
+
+	if (!strncmp(dir_start, ".in", strlen(".in"))) {
+		*dir = CAP_IN;
+		return 0;
+	}
+
+	/* check for .out */
+	dir_start = dir_end - strlen(".out");
+	if (dir_start < socketname)
+		return -EINVAL;
+
+	if (!strncmp(dir_start, ".out", strlen(".out"))) {
+		*dir = CAP_OUT;
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int parse_socket_iface(char *socketname, enum capdir dir, char *iface)
+{
+	char *iface_start, *iface_end;
+
+	/* skip UCI section name */
+	iface_start = strchr(socketname, '.');
+	if (!iface_start || iface_start[1] == '\0')
+		return -EINVAL;
+
+	iface_start += 1;
+	iface_end = socketname + strlen(socketname);
+
+	if (dir == CAP_IN)
+		iface_end -= strlen(".in.sock");
```

Comment:
The way this code deals with the filename suffixes feels very fragile... it would be great it we could split there strings in a way that didn't require pointer arithmetic.